Add support for additional multiplatform targets

05nelsonm opened the pull request:
This PR includes the following modifications:

- Adds support for additional multiplatform targets to modules `kotlin-result` and `kotlin-result-coroutines`:
     - `iosArm32`
     - `tvosArm64`
     - `tvosX64`
     - `tvosSimulatorArm64`
     - `watchosArm32`
     - `watchosArm64`
     - `watchosX64`
     - `watchosX86`
     - `watchosSimulatorArm64`
- Adds support for additional multiplatform targets to module `kotlin-result` (in addition to above):
     - `androidNativeArm32`
     - `androidNativeArm64`
     - `androidNativeX64`
     - `androidNativeX86`
     - `linuxArm32Hfp`
     - `linuxArm64`
     - `linuxMips32`
     - `linuxMipsel32`
     - `mingwX86`
     - `wasm32`
- Updates GitHub Actions test/build & publishing for all targets
- Updates the `README` with new supported platforms

Closes #77

05nelsonm commented:
Builds on all platforms

![Screenshot from 2023-01-05 20-54-30](https://user-images.githubusercontent.com/44778092/210915374-e0e1bfc4-63bf-48e2-a218-6611127e41e7.png)

michaelbull commented on .github/workflows/ci.yaml:
```diff
@@ -91,11 +91,11 @@ jobs:
-      - run: ./gradlew publishLinuxX64PublicationToMavenRepository
+      - run: ./gradlew $(./gradlew tasks | grep "PublicationToMavenRepository" | grep -e "publishAndroid" -e "publishJs" -e "publishJvm" -e "publishKotlinMultiplatform" -e "publishLinux" -e "publishWasm" | cut -d ' ' -f 1)
         if: matrix.os == 'ubuntu-latest'
 
-      - run: ./gradlew publishMingwX64PublicationToMavenRepository
+      - run: ./gradlew $(./gradlew tasks | grep "PublicationToMavenRepository" | grep "publishMingw" | cut -d ' ' -f 1)
         if: matrix.os == 'windows-latest'
 
-      - run: ./gradlew publish
+      - run: ./gradlew $(./gradlew tasks | grep "PublicationToMavenRepository" | grep -e "publishIos" -e "publishMacos" -e "publishTvos" -e "publishWatchos" | cut -d ' ' -f 1)
```
I am not a fan of the complexity going on with these shell commands. If this complexity is to live anywhere it should be in the Gradle scripts themselves, such that the commands are multi-platform and not specific to a given shell implementation (in this case, bash).

05nelsonm replied:
Resolved in commit 09d3028

michaelbull commented:
I think we will need to add a disclaimer to the coroutines section explaining the limitations on build targets in comparison with the core library, as the unfortunate platform compatibility differences will trip some people up.

05nelsonm commented:
> I think we will need to add a disclaimer to the coroutines section explaining the limitations on build targets in comparison with the core library, as the unfortunate platform compatibility differences will trip some people up.

Resolved in commit 7709128

I also fixed the badges in commit fff3239 which I fudged (sorry). Thoughts?

michaelbull commented:
Looking a lot better, thanks! I'll approve and run this to see how it fares on GitHub actions. As a disclaimer I am unsure when I will be available to actually merge and deploy this to Maven Central, I am very busy at the moment - but please don't take that as me leaving the PR dormant. I also need to get around to updating all the dependencies at some point, which has been on my backlog for a while.

05nelsonm commented:
> Looking a lot better, thanks! I'll approve and run this to see how it fares on GitHub actions. As a disclaimer I am unsure when I will be available to actually merge and deploy this to Maven Central, I am very busy at the moment - but please don't take that as me leaving the PR dormant. I also need to get around to updating all the dependencies at some point, which has been on my backlog for a while.

Right on. Sounds good, thanks!

05nelsonm commented:
Heyo, wanted to just note that Kotlin `1.8.0` added a new target; `watchosDeviceArm64`. Since coroutines hasn't had a release in forever, would simply need to:
 - Add it to the `kotlin-result` module, linking it to the intermediate source set `watchosMain` (and `watchosTest` if necessary)
 - Update CI publishing task with `publishWatchosDeviceArm64PublicationToMavenRepository`

michaelbull commented:
@05nelsonm I've squashed and rebased the changes onto master in https://github.com/michaelbull/kotlin-result/commit/6f86d20d53adaa814ca68cbb91946abcf7be2c2a

Could you please try to depend on the latest SNAPSHOT build and see if the artifacts you're expecting for each platform are consumable?

05nelsonm commented:
> @05nelsonm I've squashed and rebased the changes onto master in [6f86d20](https://github.com/michaelbull/kotlin-result/commit/6f86d20d53adaa814ca68cbb91946abcf7be2c2a)
> 
> Could you please try to depend on the latest SNAPSHOT build and see if the artifacts you're expecting for each platform are consumable?

:+1: 

![image](https://user-images.githubusercontent.com/44778092/211328835-a2551fc9-5c40-4667-9780-e30d4a223384.png)

michaelbull commented:
Thanks @05nelsonm. I'll look at the other PR during the week and also try to update dependencies, then see if I can make a release maybe on the weekend - but at least you should be able to depend upon the SNAPSHOT in the mean time.

michaelbull commented:
Released in v1.1.17.
